File: frontend/src/api/hubspot.ts
import { apiRequest } from "./client";

export type HubSpotStatus = {
  status: string;
  connected: boolean;
  portal_id: string | null;
  scopes: string[];
  created_at: string | null;
  updated_at: string | null;
  expires_at: string | null;
};

export type HubSpotContact = {
  provider_contact_id: string;
  firstname: string | null;
  lastname: string | null;
  phone: string | null;
  email: string | null;
  website: string | null;
};

type ContactPage = { contacts: HubSpotContact[]; next_after: string | null };
export type ImportResult = { imported: number; created: number; updated: number };

export const hubspotApi = {
  status: () => apiRequest<HubSpotStatus>("/api/v1/hubspot/status"),
  connect: () =>
    apiRequest<{ authorization_url: string; expires_in: number }>("/api/v1/hubspot/connect"),
  contacts: (after?: string) =>
    apiRequest<ContactPage>(`/api/v1/hubspot/contacts${after ? `?after=${encodeURIComponent(after)}` : ""}`),
  importContacts: (hubspotContactIds: string[], selectAll: boolean) =>
    apiRequest<ImportResult>("/api/v1/hubspot/import", {
      method: "POST",
      body: JSON.stringify({ hubspot_contact_ids: hubspotContactIds, select_all: selectAll }),
    }),
};

File: backend/app/schemas/hubspot.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class HubSpotConnectResponse(BaseModel):
    authorization_url: str
    expires_in: int


class HubSpotConnectionStatus(BaseModel):
    status: str
    connected: bool
    portal_id: str | None = None
    scopes: list[str] = Field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None
    expires_at: datetime | None = None


class NormalizedHubSpotContact(BaseModel):
    provider_contact_id: str
    firstname: str | None = None
    lastname: str | None = None
    phone: str | None = None
    email: str | None = None
    website: str | None = None


class HubSpotContactPage(BaseModel):
    contacts: list[NormalizedHubSpotContact]
    next_after: str | None = None


class HubSpotImportRequest(BaseModel):
    hubspot_contact_ids: list[str] = Field(default_factory=list)
    select_all: bool = False

    @model_validator(mode="after")
    def exactly_one_selection_mode(self):
        has_ids = bool(self.hubspot_contact_ids)
        if has_ids == self.select_all:
            raise ValueError("Choose contact IDs or select_all, but not both")
        self.hubspot_contact_ids = list(dict.fromkeys(self.hubspot_contact_ids))
        return self


class HubSpotImportResponse(BaseModel):
    imported: int
    created: int
    updated: int
